### PR TITLE
Async/await

### PR DESCRIPTION
**Pull request by adam-fowler**
Related Soto PR is [here](https://github.com/soto-project/soto/pull/488)

Async/Await additions
- Added async versions `AWSClient.execute`
- Added `PaginatorSequence`. Unfortunately `AsyncSequence` functions like `map`, `reduce` etc are not available. Added my own versions of `reduce`
- Added async versions of `signURL` and `signHeaders`
- Temporarily added `EventLoopFuture.get` and `EventLoopPromise.completeWithAsync` until they appear in a NIO release
- Added `AsyncCredentialProvider`
- Added Async tests

Last tested with 2021-09-15 snapshot of Swift 5.5 compiler

### CODE REVIEW

File: Sources/SotoCore/AWSClient+Paginate+async.swift
Comment (eneko) [inline Sources/SotoCore/AWSClient+Paginate+async.swift:15]: This is nice. How/where is `$AsyncAwait` defined?

**Review by eneko — COMMENTED** [review]
This is looking really good, @adam-fowler , nice work.

(I haven't tried the code myself, only reviewed it here)

Is this ready to be merged in `main` or are we waiting for more changes?

**Comment by adam-fowler** [comment]
@eneko I think this is ready to merge. Although I would prefer that AsyncSequence was more solid before merging. Currently all AsyncSequence functions are rethrows and are waiting on a swift evolution proposal for marking functions that call throwing protocol requirements rethrows instead of throws. 

Also, I'm not sure about having code dependent on unreleased swift features in the main branch. Might have a chat with swift guys about this.

File: Sources/SotoCore/AWSClient+Paginate+async.swift
Comment (adam-fowler) [reply Sources/SotoCore/AWSClient+Paginate+async.swift:15]: I just copied this from swift nio. Not sure where $AsyncAwait comes from exactly.

**Comment by eneko** [comment]
Sounds reasonable to keep the PR open until the updates are released by the Swift team

**Comment by ktoso** [comment]
Yeah I recommend holding off with the async sequence things _yet_, it might wobble around a bit in it's exact shape still.

File: Package.swift
Comment (0xTim) [inline Package.swift:28]: Any particular reason to this being `upToNextMajor`? I seem to remember it moving to `from` in a previous PR. Either way the whole manifest should be consistent

**Review by 0xTim — APPROVED** [review]
LGTM! Once AHC gets its async APIs merged in we should probably migrate the async APIs to use that to ensure tracing and logging work

File: Package.swift
Comment (adam-fowler) [reply Package.swift:28]: They were originally all `upToNextMajor`. I converted to `from` as it seemed more common. There is a possibility I did this after the async-await branch was created and then each time I would rebase with main it would jump about. Will revert to `from`

**Comment by adam-fowler** [comment]
> LGTM! Once AHC gets its async APIs merged in we should probably migrate the async APIs to use that to ensure tracing and logging work

Unfortunately that's a breaking change. The http client protocol which is public includes functions that return `EventLoopFutures`. Slightly frustrating as basically nobody cares about writing their own client. Thought maybe someone might write one with `URLSession` for iOS. 

I could make the protocol disappear in the Smithy major update and the do as you suggested.

**Pull request merged**
